tools/astarte_device_fleet_simulator/lib/astarte_device_fleet_simulator/scheduler.ex

```diff
@@ -0,0 +1,100 @@
+#
+# This file is part of Astarte.
+#
+# Copyright 2021 Ispirata Srl
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+defmodule AstarteDeviceFleetSimulator.Scheduler do
+  use GenServer, restart: :transient
+
+  require Logger
+
+  alias AstarteDeviceFleetSimulator.Device
+  alias AstarteDeviceFleetSimulator.Config
+
+  def start_link(args) do
+    GenServer.start_link(__MODULE__, args, name: __MODULE__)
+  end
+
+  @impl true
+  def init(_args) do
+    {:ok, Enum.into(Config.scheduler_opts(), %{}), {:continue, :init_messages}}
+  end
+
+  @impl true
+  def handle_continue(:init_messages, state) do
+    # not waiting all devices
+    if state.allow_messages_while_spawning do
+      check_scheduler_opts(state)
+
+      Process.send_after(self(), :terminate, state.test_duration_s)
+    end
+
+    Logger.info("Begin device spawn.")
+    Process.send(self(), :spawn, [])
+
+    {:noreply, state}
+  end
+
+  @impl true
+  def handle_cast(:device_spawn_end, state) do
+    Logger.info("Device spawn ended.")
+
+    # waiting all devices
+    if not state.allow_messages_while_spawning do
+      Process.send_after(self(), :terminate, state.test_duration_s)
+
+      Registry.dispatch(AstarteDeviceFleetSimulator.Registry, "device", fn entries ->
+        for {pid, _} <- entries, do: :gen_statem.cast(pid, :begin_publishing)
```

Comment:
Use `Enum.each` when you want to iterate on a collection and you're only interested in side effects (and not mapping the result)